torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1949,6 +1948,19 @@ void ONNXShapeTypeInference(
   SetGraphInputTypeReliable(n->owningGraph());
   GRAPH_UPDATE(
       "Running ONNX shape inference for node: ", n->kind().toDisplayString());
+
+  // Go through every output to check if they all have shape
+  // If they all do, this should be reliable even if the Op is not from ONNX.
+  size_t all_has_value = 0;
+  for (auto node_output : n->outputs()) {
+    if (auto output_type = node_output->type()->cast<TensorType>()) {
+      if (output_type->sizes().concrete_sizes().has_value()) {
```

Comment:
@BowenBao Is there a more elegant way to approach what I wanted to do here? Basically, a non-ONNX op which has shape/type information should be get in this block.

Reply:
I guess the `non-ONNX` part should be checked via looking at domain?
`concrete_sizes` might be too strict. IIRC it doesn't count if you have symbolic shape. I think you only need to differentiate from the default "empty" tensor type? I wonder if there are direct apis, but if not perhaps you can check by if rank is known.

It is important that this happens before merging shape from tracer, and after calling `setType`.
If the code is like `g.op(...).setType(...)`. I'm afraid both above conditions are not satisifed...

Reply:
Like this?
```suggestion
      if (output_type->dim()) {
```

Reply:
https://github.com/pytorch/pytorch/pull/88622#discussion_r1025891348